admin/securedrop_admin/__init__.py

```diff
@@ -544,10 +548,68 @@ def sdconfig(args):
     return 0
 
 
+def generate_new_v3_keys():
+    """This function generate new keys and returns them as tuple.
+
+    :returns: Tuple(public_key, private_key)
+    """
+
+    private_key = x25519.X25519PrivateKey.generate()
+    private_bytes = private_key.private_bytes(
+        encoding=serialization.Encoding.Raw	,
+        format=serialization.PrivateFormat.Raw,
+        encryption_algorithm=serialization.NoEncryption())
+    public_key = private_key.public_key()
+    public_bytes = public_key.public_bytes(
+        encoding=serialization.Encoding.Raw,
+        format=serialization.PublicFormat.Raw)
+
+    public = base64.b32encode(public_bytes)[:-4].decode("utf-8")
+    private = base64.b32encode(private_bytes)[:-4].decode("utf-8")
+    return public, private
+
+
+def get_v3_keys(filepath):
+    """
+    Returns the stored v3 public and private keys as Tuple.
+
+    :returns: Tuple(public_key, private_key)
+    """
+    with open(filepath) as fobj:
+        data = json.load(fobj)
+
+    return data
+
+
+def find_or_generate_new_torv3_keys(args):
+    """
+    This method will either read the old keys or generate a new
+    public/private key pair.
+    """
+    secret_key_path = os.path.join(args.ansible_path,
+                                   "tor_v3_keys.json")
```

Comment:
Just as we currently have two discrete files on the Admin Workstation for the v2 auth info, it'd be great to have the same for the v3 info. Doing so would simplify the lookup logic required in the dynamic-inventory script (not yet implemented in this PR), as well. Perhaps the function should receive `hostname=None`, and select the files that way.

Reply:
I feel that one file for admin workstation with all 3 different key pairs (journalist, ssh app, and ssh mon) and a file for journalist workstations (only with the journalist keypair) is simpler. Also, this is only for the key information, if we want, we can also remove the `ths` files all together and just the `.onion` address from the JSON files iteself (we have to add them there).

Reply:
Having taken a closer look at the keygen logic, we're still going to need separate files for the dynamic-inventory support (OK to defer to #4629). With the v2 logic, if-and-only-if the `-ths` files are present on the Admin Workstation, then Onion URLs are used for the SSH connection. Since those `-ths` files are created on the Admin Workstation at the _end_ of the install process, the logic is sound. For the v3 keygen logic, however, we need a way to denote whether the corresponding Onion URLs should be used for the SSH connection. 

I suggest moving the keygen logic into Ansible via a script, so that the JSON file does not exist when Ansible builds the SSH connection, meaning local IPv4 addresses (or v2 Onion URLs) will be used. At the end of the run, we can sprinkle in appropriate `-v3-ths` files back onto the Admin Workstation.

We cannot relegate the v3 keygen logic solely to the `securedrop-admin` init, because that's only relevant for Tails environments, which means developer staging VMs will _not_ have v3 connectivity.